RVYA25 profile (#210)
riscv/riscv-profiles: src/rvya25-profile.adoc

- *Ziccamoa* Main memory regions with both the cacheability and coherence PMAs must support all atomics in A.
- *Zicclsm* Misaligned loads and stores to main memory regions with both the
  cacheability and coherence PMAs must be supported.
- *Za64rs* Reservation sets are contiguous, naturally aligned, and a
   maximum of 64 bytes.
- *Zihintpause* Pause hint.
- *Zic64b* Cache blocks must be 64 bytes in size, naturally aligned in the
address space.
- *Zicbom* Cache-block management instructions.
- *Zicbop* Cache-block prefetch instructions.
- *Zicboz* Cache-Block Zero Instructions.
- *Zfhmin* Half-precision floating-point.
- *Zkt* Data-independent execution latency.
- *V* Vector extension.
- *Zvfhmin* Vector minimal half-precision floating-point.
- *Zvbb* Vector basic bit-manipulation instructions.
- *Zvkt* Vector data-independent execution latency.
- *Zihintntl* Non-temporal locality hints.
- *Zicond* Integer conditional operations.
- *Zimop* may-be-operations.
- *Zcmop* Compressed may-be-operations.
- *Zcb* Additional compressed instructions.
- *Zfa* Additional floating-Point instructions.
- *Zawrs* Wait-on-reservation-set instructions.

> The following mandatory extensions are RVY specific, but may be replaced over time with binary compatible alternatives if better schemes are developed:
> 
> - *Zys* Sentry definition
> - *Zylevels* Capability flow control
> 
> The following mandatory extension is in addition to RVA23U64:

Author: ```suggestion

The following mandatory extension is RVY specific and specifies the capability encoding format:

-* Zydefaultcap* Default capability encoding (visible to software as uycfg.base=0).

The following mandatory extensions are RVY specific, but may be replaced over time with binary compatible alternatives if better schemes are developed:

- *Zys* Sentry definition
- *Zylevels1* 2-level Capability Information Flow Control (visible to software as uycfg.features=1).

The following mandatory extension is in addition to RVA23U64:

```

Author: @nwf any good?

Reviewer: I didn't think we were using extensions to capture encodings?  If that's fine with RVI, I think I'm happy to roll with it.

- *Supm* Pointer masking, with the execution environment providing a means to
   select PMLEN=0 and PMLEN=7 at minimum.

NOTE: How to benefit from the combination of pointer masking and CHERI is a research topic.

==== RVYA25U64 Optional Extensions

===== Localized Options

The following localized options are taken from RVA23U64:

- *Zvkng* Vector crypto NIST algorithms with GCM.
- *Zvksg* Vector crypto ShangMi algorithms with GCM.

===== Development Options

The following are new development options intended to become mandatory in a future RVA/RVYA profile.

- *Zacas* Compare-and-Swap instructions.
- *Ziccamoc* Main memory regions with both the cacheability and coherence PMAs
  must provide `AMOCASQ` level PMA support.
- *Zvbc* Vector carryless multiplication.
- *Zama16b* Misaligned loads, stores, and AMOs to main memory regions that do not cross a naturally aligned 16-byte boundary are atomic.

===== Expansion Options

The following expansion options were also present in RVA22U64:

- *Zfh* Scalar half-precision floating-point.
- *Zbc* Scalar carryless multiply.
- *Zvfh* Vector half-precision floating-point.
- *Zfbfmin* Scalar BF16 converts.
- *Zvfbfmin* Vector BF16 converts.

> - *Zvfbfwma* Vector BF16 widening mul-add.
> 
> NOTE: *Zicfilp* Landing Pads, and *Zicfiss* Shadow Stack are redundant for CHERI, and so are currently listed as incompatible.

Author: ```suggestion
- *Zvfbfwma* Vector BF16 widening mul-add.
- *Zicfilp* Landing Pads

 NOTE: *Zicfiss* Shadow Stack is incompatible with RVY.
```

- *Zicbop* Cache-block prefetch instructions.
- *Zicboz* Cache-Block Zero Instructions.
- *Zfhmin* Half-precision floating-point.
- *Zkt* Data-independent execution latency.
- *V* Vector extension.
- *Zvfhmin* Vector minimal half-precision floating-point.
- *Zvbb* Vector basic bit-manipulation instructions.
- *Zvkt* Vector data-independent execution latency.
- *Zihintntl* Non-temporal locality hints.
- *Zicond* Integer conditional operations.
- *Zimop* may-be-operations.
- *Zcmop* Compressed may-be-operations.
- *Zcb* Additional compressed instructions.
- *Zfa* Additional floating-Point instructions.
- *Zawrs* Wait-on-reservation-set instructions.

The following mandatory extension is RVY specific:

- *Zys* Sentry definition

The following mandatory extension is in addition to RVA23U64:

- *Zabhlrsc* Byte and halfword load-reserved and store conditional

The following development option from RVA23U645 is mandatory in RVYA25U64:

- *Zabha* Byte and halfword atomic memory operations.

The following extension is mandatory in RVA23 but is _incompatible_ with RVY:

- *Supm* Pointer masking, with the execution environment providing a means to
   select PMLEN=0 and PMLEN=7 at minimum.

> NOTE: How to benefit from the combination of pointer masking and CHERI is a research topic.

Author: ```suggestion
NOTE: How to benefit from the combination of pointer masking and CHERI is a research topic.
 RVY harts which also implement Zyhybrid can make use of pointer masking.
```

- *Sstvecd* `stvec.MODE` must be capable of holding the value 0
  (Direct).  When `stvec.MODE=Direct`, `stvec.BASE` must be capable of
  holding any valid four-byte-aligned address.

- *Sstvala* `stval` must be written with the faulting virtual address
  for load, store, and instruction page-fault, access-fault, and
  misaligned exceptions, and for breakpoint exceptions other than
  those caused by execution of the `EBREAK` or `C.EBREAK` instructions.
  For virtual-instruction and illegal-instruction exceptions, `stval` must be written with the
  faulting instruction.

- *Sscounterenw* For any `hpmcounter` that is not read-only zero, the
  corresponding bit in `scounteren` must be writable.

- *Svpbmt* Page-based memory types

- *Svinval* Fine-grained address-translation cache invalidation.

- *Svnapot* NAPOT translation contiguity.

- *Sstc* supervisor-mode timer interrupts.

- *Sscofpmf* count overflow and mode-based filtering.

- *Ssu64xl* `sstatus.UXL` must be capable of holding the value 2
(i.e., UXLEN=64 must be supported).

- *Sha* The augmented hypervisor extension.

NOTE: Sha is defined in the RVA23 profile document.

> The following mandatory extension is RVY25S64 specific:
> 
> - *Svy* Read/write access control for capability tags in the page table entries.

Author: ```suggestion
```

===== Expansion Options

- *Sv48* Page-based 48-bit virtual-memory system.

- *Sv57* Page-based 57-bit virtual-memory system.

- *Zkr*  Entropy CSR.

- *Svadu* Hardware A/D bit updates, and RVY revocation updates if Svucrg is implemented.

- *Sdtrig* Debug triggers.

- *Ssstrict* No non-conforming extensions are present.  Attempts to
   execute unimplemented opcodes or access unimplemented CSRs in the
   standard or reserved encoding spaces raises an illegal instruction
   exception that results in a contained trap to the supervisor-mode
   trap handler.

NOTE: The RVA23 profile contains more details of Ssstrict.

- *Svvptc* Transitions from invalid to valid PTEs will be visible in
   bounded time without an explicit memory-management fence.

The following extension is an expansion option in RVA23 but is _incompatible_ with RVY:

- *Sspm* Supervisor-mode pointer masking, with the supervisor execution
   environment providing a means to select PMLEN=0 and PMLEN=7 at minimum.

NOTE: How to benefit from the combination of pointer masking and RVY is a research topic.

===== Transitory Options

> The RYA25S64 hybrid options are transitory as over time it is expected that the compatibility requirement will disappear.
> 
> - *Smyre* Support for enabling CHERI at different privilege levels (implies Zyhybrid)
> 
> NOTE: Implementing *Smyre* requires that the implementation supports the full RVA23S64 profile.

Author: ```suggestion
The RYA25S64 hybrid options are transitory as over time it is expected that the compatibility requirement will disappear.

- *Zyhybrid for privileged architectures* Support for enabling CHERI at different privilege levels (implies, and implied by Zyhybrid)
```

discoverable coarse-grain options.  It is explicitly a non-goal of
RVYA25 to allow more hardware implementation flexibility by supporting
only a minimal set of features and a large number of fine-grain
extensions.

Only user-mode (RVYA25U64) and supervisor-mode (RVYA25S64) profiles are
specified in this family.

NOTE: This profile is presented as incremental changes over RVA23.

=== RVYA25U64 Profile

The RVYA25U64 profile specifies the ISA features available to user-mode
execution environments in 64-bit applications processors.  This is the
most important profile within the application processor family in
terms of the amount of software that targets this profile.

==== RVYA25U64 Mandatory Base

RVY64I is the mandatory base ISA for RVYA25U64 and is little-endian.  As
per the unprivileged architecture specification, the `ECALL`
instruction causes a requested trap to the execution environment.

NOTE: The `ECALL` requirement is as specified in RVA23U64.

==== RVYA25U64 Mandatory Extensions

The following mandatory extensions are also mandatory in RVA23U64.

- *M* Integer multiplication and division.
- *A* Atomic instructions.
- *F* Single-precision floating-point instructions.

> - *D* Double-precision floating-point instructions.
> - *B* Bit-manipulation instructions.

Author: ```suggestion
- *D* Double-precision floating-point instructions.
- *C* Compressed instructions.
- *B* Bit-manipulation instructions.
```

- *Ziccamoa* Main memory regions with both the cacheability and coherence PMAs must support all atomics in A.
- *Zicclsm* Misaligned loads and stores to main memory regions with both the
  cacheability and coherence PMAs must be supported.
- *Za64rs* Reservation sets are contiguous, naturally aligned, and a
   maximum of 64 bytes.
- *Zihintpause* Pause hint.
- *Zic64b* Cache blocks must be 64 bytes in size, naturally aligned in the
address space.
- *Zicbom* Cache-block management instructions.
- *Zicbop* Cache-block prefetch instructions.
- *Zicboz* Cache-Block Zero Instructions.
- *Zfhmin* Half-precision floating-point.
- *Zkt* Data-independent execution latency.
- *V* Vector extension.
- *Zvfhmin* Vector minimal half-precision floating-point.
- *Zvbb* Vector basic bit-manipulation instructions.
- *Zvkt* Vector data-independent execution latency.
- *Zihintntl* Non-temporal locality hints.
- *Zicond* Integer conditional operations.
- *Zimop* may-be-operations.
- *Zcmop* Compressed may-be-operations.
- *Zcb* Additional compressed instructions.
- *Zfa* Additional floating-Point instructions.
- *Zawrs* Wait-on-reservation-set instructions.

> NOTE: The combination of "C" and "D" is incompatible with RVY as Zcd is an incompatible extension.

Reviewer 2: Zcd can still be supported for Zyhybrid?

Reviewer 3: Also, C should just mean "the subset of C that is valid for RVY". Making RV64YC be illegal is just going to be obnoxious to deal with.

Author: >Zcd can still be supported for Zyhybrid?

Of course, in that mode you get full RVA23 (I write that further down in the profile). It's the remapping of C.FLD->C.LY etc.

>Also, C should just mean "the subset of C that is valid for RVY". Making RV64YC be illegal is just going to be obnoxious to deal with.

ok - fair point - and this affects the spec as well as the profile - what you're saying is that C + D on RVY would _not_ include Zcd, you'd get C.LY[SP], C.SY[SP].

that works, but we then need to put an extension name back for C.LY[SP], C.SY[SP] so we can say that we get Zca + Zclsy (or whatever we call it). I agree it seems sensible to be able to specify C on the command line.

Author: https://github.com/riscv/riscv-cheri/pull/762

Author: done.

> 
> Therefore the following mandatory extension is included instead:
> 
> - *Zca* Compressed encodings excluding floating point loads and stores, including capability loads and stores for RVY

Author: ```suggestion
```

The following mandatory extension is RVY specific:

- *Zys* Sentry definition

The following mandatory extension is in addition to RVA23U64:

- *Zabhlrsc* Byte and halfword load-reserved and store conditional

The following development option from RVA23U645 is mandatory in RVYA25U64:

- *Zabha* Byte and halfword atomic memory operations.

> The following extension is mandatory in RVA23 but is _incompatible_ with RVY:

Reviewer 3: It's fine for integer mode code? Just mask the address and let DDC authorise that.

   select PMLEN=0 and PMLEN=7 at minimum.

NOTE: How to benefit from the combination of pointer masking and CHERI is a research topic.

==== RVYA25U64 Optional Extensions

===== Localized Options

The following localized options are taken from RVA23U64:

- *Zvkng* Vector crypto NIST algorithms with GCM.
- *Zvksg* Vector crypto ShangMi algorithms with GCM.

===== Development Options

The following are new development options intended to become mandatory in a future RVA/RVYA profile.

- *Zacas* Compare-and-Swap instructions.
- *Ziccamoc* Main memory regions with both the cacheability and coherence PMAs
  must provide `AMOCASQ` level PMA support.
- *Zvbc* Vector carryless multiplication.
- *Zama16b* Misaligned loads, stores, and AMOs to main memory regions that do not cross a naturally aligned 16-byte boundary are atomic.

===== Expansion Options

The following expansion options were also present in RVA22U64:

- *Zfh* Scalar half-precision floating-point.
- *Zbc* Scalar carryless multiply.
- *Zvfh* Vector half-precision floating-point.
- *Zfbfmin* Scalar BF16 converts.
- *Zvfbfmin* Vector BF16 converts.
- *Zvfbfwma* Vector BF16 widening mul-add.

> NOTE: *Zicfilp* Landing Pads, and *Zicfiss* Shadow Stack are redundant for CHERI, and so are currently listed as incompatible.

Reviewer 3: As above. You want to support them for integer mode code, and Zicfilp has no incompatibility with CHERI. Yes, it may not be needed, but it isn't incompatible, and it can be complementary.

Reviewer 3: Zicfiss you say integer only and doesn't exist for RVY.

Zicfilp I see no problem with allowing, especially given the labelling scheme has the ability to do some type enforcement that CHERI isn't designed to do, rather than just prevent JOP which we already do.

Author: OK - let's go with that for now.

  (Direct).  When `stvec.MODE=Direct`, `stvec.BASE` must be capable of
  holding any valid four-byte-aligned address.

- *Sstvala* `stval` must be written with the faulting virtual address
  for load, store, and instruction page-fault, access-fault, and
  misaligned exceptions, and for breakpoint exceptions other than
  those caused by execution of the `EBREAK` or `C.EBREAK` instructions.
  For virtual-instruction and illegal-instruction exceptions, `stval` must be written with the
  faulting instruction.

- *Sscounterenw* For any `hpmcounter` that is not read-only zero, the
  corresponding bit in `scounteren` must be writable.

- *Svpbmt* Page-based memory types

- *Svinval* Fine-grained address-translation cache invalidation.

- *Svnapot* NAPOT translation contiguity.

- *Sstc* supervisor-mode timer interrupts.

- *Sscofpmf* count overflow and mode-based filtering.

- *Ssu64xl* `sstatus.UXL` must be capable of holding the value 2
(i.e., UXLEN=64 must be supported).

- *Sha* The augmented hypervisor extension.

NOTE: Sha is defined in the RVA23 profile document.

The following mandatory extension is RVY25S64 specific:

> - *Svy* Read/write access control for capability tags in the page table entries.

Reviewer 3: I thought we weren't inventing new names for things that are part of the base ISA when S-mode is present?

Reviewer 2: I believe we've removed all the other ones, but this one I am not sure if ARC wants an explicit name since it does not apply to all translation schemes (Sv32). I will explicitly ask about this one when I send the review email.

Reviewer 3: I don't think it's helpful to frame something as a separate extension when it's not separable from the thing it's extending

The following extension is mandatory in RVA23 but is _incompatible_ with RVY:

> - *Ssnpm* Pointer masking, with `senvcfg.PMM` and `henvcfg.PMM` supporting,

Reviewer 3: See Sunpm

at minimum, settings PMLEN=0 and PMLEN=7.

NOTE: How to benefit from the combination of pointer masking and CHERI is a research topic.

==== RVYA25S64 Optional Extensions

===== Localized Options

There are no privileged localized options in RVYA25S64.

===== Development Options

There privileged development options in RVYA25S64.

> - *Svucrg* Page-based revocation (preventing use-after-free in virtual memory systems).

Author: I've got it as mandatory but in the development options and intend to do the same with the sentry extension.
We all agree that both are mandatory but may change over time.
I'll ask about the correct approach here (I'll email the TSC).

> - *Svucrglct* (implies *Svucrg*) checking loaded capability tags before taking exceptions to optimize revocation.

Reviewer 3: I remain unconvinced that having this be optional is a good idea, but that's a separate issue for the RVY spec itself about whether this extension should even exist or just be part of Svucrg

Author: we had a useful discussion about it today, @nwf pionted out that if you support synchronous ECC errors on loaded data then you should support the tag checking version, and if not, then you supoprt the simpler version which doesn't check tags.

- *Sv57* Page-based 57-bit virtual-memory system.

- *Zkr*  Entropy CSR.

- *Svadu* Hardware A/D bit updates, and RVY revocation updates if Svucrg is implemented.

- *Sdtrig* Debug triggers.

- *Ssstrict* No non-conforming extensions are present.  Attempts to
   execute unimplemented opcodes or access unimplemented CSRs in the
   standard or reserved encoding spaces raises an illegal instruction
   exception that results in a contained trap to the supervisor-mode
   trap handler.

NOTE: The RVA23 profile contains more details of Ssstrict.

- *Svvptc* Transitions from invalid to valid PTEs will be visible in
   bounded time without an explicit memory-management fence.

The following extension is an expansion option in RVA23 but is _incompatible_ with RVY:

- *Sspm* Supervisor-mode pointer masking, with the supervisor execution
   environment providing a means to select PMLEN=0 and PMLEN=7 at minimum.

NOTE: How to benefit from the combination of pointer masking and RVY is a research topic.

===== Transitory Options

The RYA25S64 hybrid options are transitory as over time it is expected that the compatibility requirement will disappear.

> - *Smyre* Support for enabling CHERI at different privilege levels (implies Zyhbrid)

Reviewer 3: This should just be part of having Zyhybrid (which isn't spelled correctly here) and modes other than M-mode

Author: see https://github.com/riscv/riscv-cheri/pull/763

NOTE: Implementing *Smyre* requires that the implementation supports the full RVA23S64 profile.

==== RVYA25S64 Recommendations

- Implementations are strongly recommended to raise
  illegal-instruction exceptions when attempting to execute
  unimplemented opcodes or access unimplemented CSRs.

> - Implementions which support Zyhbrid are strongly recommended

Reviewer 3: Required

Author: if we merge them, then certainly!

Author: https://github.com/riscv/riscv-cheri/pull/763

 to support Smyre.

== Glossary of ISA Extensions

The following unprivileged (and unratified) ISA extension is RVY specific:

- *Zys* Sentry definition
- *Zyhybrid* Support for RV64I base architecture

> The following unprivileged (and unratified) ISA extension is generalpurpose but added for CHERI software:

Reviewer 2: ```suggestion
The following unprivileged (and unratified) ISA extension is general-purpose but added for CHERI software:
```
